src/main/scala/chisel3/util/MemInterface.scala

```diff
@@ -0,0 +1,263 @@
+  val enable = Input(Bool())
+  val data = Output(tpe)
+  val clock = Input(Clock())
+}
+
+/** A bundle of signals representing a memory write port.
+  *
+  * @tparam tpe The data type of the memory port
+  * @param addrWidth The width of the address signal
+  * @param masked Whether this read/write port should have an optional mask.
+  *
+  * @note `masked` is only valid if tpe is a Vec; if this is not the case no mask will be initialized
+  *       regardless of the value of `masked`.
+  */
+class MemoryWritePort[T <: Data] private[chisel3] (tpe: T, addrWidth: Int, masked: Boolean) extends Bundle {
+  val address = Input(UInt(addrWidth.W))
+  val enable = Input(Bool())
+  val data = Input(tpe)
+  val mask: Option[Vec[Bool]] = if (masked) {
+    val maskSize = tpe match {
+      case vec: Vec[_] => vec.size
+      case _ => 0
+    }
+    Some(Input(Vec(maskSize, Bool())))
+  } else {
+    None
+  }
+  val clock = Input(Clock())
+}
+
+/** A bundle of signals representing a memory read/write port.
+  *
+  * @tparam tpe The data type of the memory port
+  * @param addrWidth The width of the address signal
+  * @param masked Whether this read/write port should have an optional mask.
+  *
+  * @note `masked` is only valid if tpe is a Vec; if this is not the case no mask will be initialized
+  *       regardless of the value of `masked`.
+  */
+class MemoryReadWritePort[T <: Data] private[chisel3] (tpe: T, addrWidth: Int, masked: Boolean) extends Bundle {
+  val address = Input(UInt(addrWidth.W))
+  val enable = Input(Bool())
+  val isWrite = Input(Bool())
+  val readData = Output(tpe)
+  val writeData = Input(tpe)
+  val mask: Option[Vec[Bool]] = if (masked) {
+    val maskSize = tpe match {
+      case vec: Vec[_] => vec.size
+      case _ => 0
+    }
+    Some(Input(Vec(maskSize, Bool())))
+  } else {
+    None
+  }
+  val clock = Input(Clock())
+}
+
+/** A IO bundle of signals connecting to the ports of a wrapped `SyncReadMem`, as requested by
+  * `SRAMInterface.apply`.
+  *
+  * @tparam tpe The data type of the memory port
+  * @param width The width of the address wires of each port
+  * @param numRd The number of read ports
+  * @param numWr The number of write ports
+  * @param numRdWr The number of read/write ports
+  */
+class SRAMInterface[T <: Data](tpe: T, addrWidth: Int, numRd: Int, numWr: Int, numRdWr: Int, masked: Boolean)
+    extends Bundle {
+  if (masked) {
+    require(
+      tpe.isInstanceOf[Vec[_]],
+      s"masked writes require that SRAMInterface is instantiated with a data type of Vec (got $tpe instead)"
+    )
+  }
+  override def typeName: String = s"SRAMInterface_${SRAM.portedness(numRd, numWr, numRdWr)}"
+
+  val rd: Vec[MemoryReadPort[T]] = Vec(numRd, new MemoryReadPort(tpe, addrWidth))
+  val wr: Vec[MemoryWritePort[T]] = Vec(numWr, new MemoryWritePort(tpe, addrWidth, masked))
+  val rw: Vec[MemoryReadWritePort[T]] = Vec(numRdWr, new MemoryReadWritePort(tpe, addrWidth, masked))
```

Comment:
I still don't like these names since they are extremely abbreviated whereas we are not using abbreviations in the fields of those actual Bundles. My two proposals are:
* `reads`, `writes`, and `readwrites`
* `readPorts`, `writePorts`, and `readwritePorts`.

Reply:
> readPorts, writePorts, and readwritePorts.
+1